rs/tests/networking/read_state_test.rs

```diff
@@ -809,6 +841,103 @@ fn test_subnet_canister_ranges_paths(env: TestEnv, version: read_state::subnet::
     validate_canister_ranges(&subnet, &path, &cert);
 }
 
+/// Checks that requesting the deprecated canister ranges path (/subnet/{subnet_id}/canister_ranges)
+/// doesn't work on the v3 endpoints, except when subnet_id == nns_subnet_id.
+fn test_deprecated_subnet_canister_ranges_paths(env: TestEnv, endpoint: Endpoint) {
+    let should_accept_deprecated_canister_ranges_path = match endpoint {
+        Endpoint::CanisterReadState(read_state::canister::Version::V2)
+        | Endpoint::SubnetReadState(read_state::subnet::Version::V2) => true,
+        Endpoint::CanisterReadState(read_state::canister::Version::V3)
+        | Endpoint::SubnetReadState(read_state::subnet::Version::V3) => false,
+    };
+    let app_subnet = get_first_app_subnet(&env);
+    let nns_subnet = env.topology_snapshot().root_subnet();
+
+    let deprecated_canister_ranges_for_app_subnet_path: Vec<Label<Vec<u8>>> = vec![
+        "subnet".into(),
+        app_subnet.subnet_id.get_ref().as_slice().into(),
+        "canister_ranges".into(),
+    ];
+
+    let response = read_state(
+        &env,
+        vec![deprecated_canister_ranges_for_app_subnet_path.clone()],
+        endpoint,
+    );
+
+    if should_accept_deprecated_canister_ranges_path {
+        let certificate = response.expect(
+            "The old endpoints should correctly handle the request with the deprecated paths",
+        );
+        assert!(
+            lookup_value(
+                &certificate,
+                deprecated_canister_ranges_for_app_subnet_path.clone()
+            )
+            .is_ok(),
+            "The certificate should contain the deprecated canister ranges"
+        );
+    } else {
+        let err = response.expect_err("The new endpoints should not accept the deprecated paths");
+        assert_matches!(err, AgentError::HttpError(payload) if payload.status == StatusCode::NOT_FOUND.as_u16());
+    }
+
+    let deprecated_canister_ranges_for_nns_path: Vec<Label<Vec<u8>>> = vec![
+        "subnet".into(),
+        nns_subnet.subnet_id.as_ref().as_slice().into(),
+        "canister_ranges".into(),
+    ];
+
+    let certificate = read_state(
+        &env,
+        vec![deprecated_canister_ranges_for_nns_path.clone()],
+        endpoint,
+    )
+    .expect("Requesting the deprecated canister ranges for the nns subnet is always allowed");
+    assert!(
+        lookup_value(
+            &certificate,
+            deprecated_canister_ranges_for_nns_path.clone()
+        )
+        .is_ok(),
+        "The certificate should contain the deprecated canister \
+        ranges because we are requesting the nns subnet"
+    );
+
+    for path in [
+        vec!["subnet".into()],
+        vec![
+            "subnet".into(),
+            app_subnet.subnet_id.get_ref().as_slice().into(),
+        ],
```

Comment:
Maybe we could have those tests for NNS subnet, too, to ensure that the canister ranges path is present for the NNS subnet even if not explicitly requested.